device/pci_device.cpp

```diff
@@ -0,0 +1,431 @@
+void TTDevice::open_hugepage_per_host_mem_ch(uint32_t num_host_mem_channels) {
+    for (int ch = 0; ch < num_host_mem_channels; ch++) {
+        log_debug(LogSiliconDriver, "Opening device_fd_per_host_ch device index: {} ch: {} (num_host_mem_channels: {})", device_id, ch, num_host_mem_channels);
+        int device_fd_for_host_mem = find_device(device_id);
+        if (device_fd_for_host_mem == -1) {
+            throw std::runtime_error(std::string("Failed opening a host memory device handle for device ") + std::to_string(device_id));
+        }
+        device_fd_per_host_ch.push_back(device_fd_for_host_mem);
+    }
+}
+
+tt::ARCH TTDevice::get_arch() const {
+    return arch;
+}
+
+template<typename T>
+T* TTDevice::get_register_address(std::uint32_t register_offset) {
+    void *reg_mapping;
+    if (system_reg_mapping != nullptr && register_offset >= system_reg_start_offset) {
+        register_offset -= system_reg_offset_adjust;
+        reg_mapping = system_reg_mapping;
+    } else if (bar0_wc != bar0_uc && register_offset < bar0_wc_size) {
+        reg_mapping = bar0_wc;
+    } else {
+        register_offset -= bar0_uc_offset;
+        reg_mapping = bar0_uc;
+    }
+    return reinterpret_cast<T*>(static_cast<uint8_t*>(reg_mapping) + register_offset);
+}
+
+void TTDevice::write_block(uint64_t byte_addr, uint64_t num_bytes, const uint8_t* buffer_addr) {
+    void *dest = nullptr;
+    if (bar4_wc != nullptr && byte_addr >= BAR0_BH_SIZE) {
+        byte_addr -= BAR0_BH_SIZE;
+        dest = reinterpret_cast<uint8_t *>(bar4_wc) + byte_addr;
+    }else {
+        dest = get_register_address<uint8_t>(byte_addr);
+    }
+
+    const void *src = reinterpret_cast<const void *>(buffer_addr);
+    memcpy(dest, src, num_bytes);
+// #ifndef DISABLE_ISSUE_3487_FIX
+//     // memcpy_to_device(dest, src, num_bytes);
+// #else
+//     // ~4x faster than pci_read above, but works for all sizes and alignments
+//     memcpy(dest, src, num_bytes);
+// #endif
+}
+
+void TTDevice::read_block(uint64_t byte_addr, uint64_t num_bytes, uint8_t* buffer_addr) {
+    void *src = nullptr;
+    if (bar4_wc != nullptr && byte_addr >= BAR0_BH_SIZE) { //arch == tt::ARCH::BLACKHOLE && 
+        byte_addr -= BAR0_BH_SIZE;
+        src = reinterpret_cast<uint8_t *>(bar4_wc) + byte_addr;
+    } else {
+        src = get_register_address<uint8_t>(byte_addr);
+    }
+
+    void *dest = reinterpret_cast<void *>(buffer_addr);
+    memcpy(dest, src, num_bytes);
+// #ifndef DISABLE_ISSUE_3487_FIX
+//     // memcpy_from_device(dest, src, num_bytes);
+// #else
+//     // ~4x faster than pci_read above, but works for all sizes and alignments
+//     memcpy(dest, src, num_bytes);
+// #endif
+}
+
+// This is only needed for the BH workaround in iatu_configure_peer_region since no arc
+void TTDevice::write_regs(volatile uint32_t *dest, const uint32_t *src, uint32_t word_len) {
```

Comment:
Those are doing one register at a time, so the while loop and the `word_len` aren't needed.

I am picking on this bit of code because of how it's used elsewhere - I'll leave a comment at that location.